Feature ld polygon editor

jaschutte opened the pull request:
Implements the polygon editor for the level designer. This PR aims to fully implement the polygon editor. This editor is supposed to work for all "things" which require a polygon to edit. 

## Todo list

- [x] Select polygons
- [x] Place polygons
- [x] Edit polygon vertices
- [x] Move polygon vertices
- [x] Erase polygon vertices
- [ ] Active edge selector (for water, is this edge have ripples? for terrain, should this edge have grass?)
- [ ] Proper UI/UX

jaschutte committed:
Keybinds for cancelling/finishing placement

jaschutte committed:
Groundwork for detecting new vertex insertions

jaschutte committed:
Start interaction on press

jaschutte committed:
Fixed system not crashing if there's no add vertex button

jaschutte committed:
Added new vertex support

jaschutte committed:
Reworked reparent_buttons

jaschutte committed:
Saving & discarding changes

jaschutte committed:
Safe zone for corner buttons

I'm not happy with this, maybe we should split the moving of old nodes
and placing of new ones into two seperate modes. Since right now it is
very finnicky.

jaschutte committed:
Fixed unable to drag index 0

Koopa1018 committed:
Save Polygon Editor in its own scene

For easier editing in isolation.

Koopa1018 committed:
Move vertex icons into poly-edit folder

Not great having them mixed in with everything else, especially when there's a folder for it now.

Koopa1018 committed:
Rename should_have_buttons to show_verts

The original name annoys me because the fact that verts are buttons is an implementation detail. Calling code only needs to care about whether "vertices" are visible.

Koopa1018 committed:
Heckithit shorten the other "should"s too

These names don't confuse; the word "should" just doesn't add information.

Koopa1018 committed:
Move vertex buttons into node tree

Godot docs say instantiating like this is actually faster than spawning them in code. Theoretically more maintainable too. I can't imagine the performance of this code is actually going to make a dent....

Not integrated into code yet.

Koopa1018 committed:
Rename AddVert for consistency with scripts

NewVert, not AddVert!

Koopa1018 committed:
Reconnect add-terrain button to poly-editor

This signal seems to have come detached in making the poly-editor its own scene?

Koopa1018 committed:
Use prefab placed verts

Nothing seems to have broken. That's good.

Koopa1018 committed:
Use prefab add vert

This one is weirder...when I was testing, I found that recycling the same node over and over would cause confusing behavior when adding verts. The problem didn't occur when we were deleting and reconstructing the add button every time, so...just do that now, despite the premade node apparently being perfectly good as-is?

Koopa1018 committed:
Comment reparent_buttons()

Koopa1018 committed:
Comment & cleanup edge drawing routine

Simplify predict-line drawing into a single clause of an if statement, rather than being scattered across two different spots with normal-line drawing awkwardly rolled into one of them using inline if.

Also, remove "transparent_connector," as it was perfectly synonymous with draw_predict_line and wasn't even respected during the second predict-line draw.

Koopa1018 committed:
Comment and cleanup add-vert placement

For starters, identify what this section of code even is. Then the usual comment, clarify names, de-square the constant's value for ease of editing, etc. Trying hard to match existing logic behavior.

Unintuitively, "if mouse_position" is always true. If mouse position is ever null (is that possible?), the code which calculates nearest_position will crash before the code even reaches the if statement. In other words, nearest_position implies mouse_position, so it's valid logically to only check the former.

Koopa1018 committed:
Script for buttons with separate lmb+rmb signals

Based on the stackexchange post linked in the comment. Untested, but can confirm adding it to placed-vert didn't break anything.

Koopa1018 committed:
Connect move-button signal only to left click

Well that's interesting, This event still fires when I press LMB on a vertex, move mouse off, then release. Doesn't do that with vanilla buttons--moving mouse off the button always cancels the mouse click....

Koopa1018 committed:
Ignore LMB/RMB events if mouse has moved

Now I've got the opposite problem: the button is almost never responding to anything....

Koopa1018 committed:
Properly detect RMB-button mouse-off state

Ah! Here we go. Everything works when it should, and doesn't when it shouldn't.

Koopa1018 committed:
Refactor add/move signal & callback naming

Koopa1018 committed:
Add delete-vertex signal, integrate with RMB

Doesn't call any code yet, but it's ready to, and that's a good start!

Koopa1018 committed:
Call PolygonEditor fns directly from PolygonDraw

Decoupling is all well and good, but. Are we *really* ever going to use PolygonDraw except as a child of PolygonEditor? And if not, why add the extra indirection?

In the process, refactor polygon_editor's function names to be more external-caller-friendly.

Koopa1018 committed:
Misc refactoring to the polygon editor`

...taking pains to avoid changing logic. Lots of changes to support type-hinting, and encapsulate some polygon_draw functionality behind functions so polygon_draw is in charge of more of its own logic.

(Add an assertion in the one place I didn't quite preserve logic--just to make sure we know if the weird edge case that triggers it ever happens?)

Koopa1018 committed:
Reorder functions in polygon_editor

...to standard, and also for clarity and intuitiveness.

Koopa1018 committed:
Finally get vertex removal implemented

Not yet set up to handle going below 3 vertices, but it's a step forward.

Koopa1018 committed:
Delete terrain when too few verts remain

Nearly there! Need to update the selection handler so that the move/copy/polyedit buttons don't stay when the terrain it points to is gone.

Koopa1018 committed:
Underscore private fns in selection_handler

I just need to know on sight which of these is or isn't publicly used....

Koopa1018 committed:
Formal API for de/selecting things

I can't imagine that deleting terrain within the polygon editor will be the only time we need to change selections from the outside in the future. Even if it isn't, having all the actual select-deselect code in the same place can't possibly hurt when it comes to understanding the system.

(also, add used parameters to the signature of the selection_changed signal. oh no signal signature)

Koopa1018 committed:
Heckithit this fn can be static

While I'm refactoring this file anyway.

Koopa1018 committed:
Connect polygon deletion to selection

Annd...done. Using the new selection-modification API and an on-delete signal (to avoid passing in more dependencies), polygons now properly deselect themselves when they self-destruct due to too few vertices. Vertex deletion might just be finished.

Koopa1018 committed:
Create new-vert button before placed verts

Makes it easier to select placed verts. Unfortunately, it also makes it actively hard to add verts in some cases.... is it possible to tune buttons' response distances, I wonder?

Koopa1018 committed:
Revert vertex widgets to default size

This is why the selection was acting jank! The buttons' hit areas were significantly larger below and to the right than the visuals were. Fixing that instantly makes things suck less.

Koopa1018 committed:
Different cursor shape for placed verts

Nice little bit of extra feedback. (I want to use the crossed arrows cursor, but alas, that one means "click and drag," so I can't unless/until we have that.)

Koopa1018 committed:
Reformat _unhandled_input as a match statement

Woot! Instant clarity boost.

Koopa1018 committed:
Refactor polyedit _process to use refresh

...instead of doing a weird, apparently useless identity assign to trigger the setter.

Koopa1018 committed:
Finish commenting polygon editor/draw

.

Koopa1018 committed:
Comment out adding extra close-gap vertex

As the code comment says: it's not really doing anything to fix the terrain, and it's causing usability problems where vertex 0 separates into two when you click it (and it's easier doing this than figuring out how to keep two verts glued together in the editor).

Koopa1018 committed:
Make in_level a formal EDITOR_STATE

...instead of a free-floating bool. This ensures state-exit code (which might e.g. ensure state-specific widgets get cleaned up) runs before playtesting begins.

Incidentally, move the start button's state change call before any of the other logic, to ensure the state-change signal emits while the editor is in a fully valid state.

Koopa1018 committed:
Formalize signature of editor_state_changed

Koopa1018 committed:
Revert "Make in_level a formal EDITOR_STATE"

This reverts commit 0227714ac6f7add698071f902a70f09d05ff661e.

I'm not convinced this was a good idea, and it certainly wasn't a good solution to the problem I'm trying to fix.

Koopa1018 committed:
Add playtest start/end signals to main

It feels like Main should be responsible for playtest start/end somehow? It kind of already is with the "ld_exit" check.

These signals will make it much easier to keep poly-edit out of inconsistent states.

Koopa1018 committed:
Finalize polygon edits when playtest starts

Before this commit, playtesting while polygon-editing would leave the edited polygon in a weird, invisible-but-still-solid state, and with its pre-edit geometry. No longer--pressing play now commits the terrain edits the same as pressing ld_poly_finish.

Koopa1018 committed:
Emit playtest_ended fully after end-playtest

To parallel playtest_started in guaranteeing a fully valid LD UI. Actually--document that guarantee so calling code knows it can rely on it.

Koopa1018 committed:
Relocate deselect-on-delete logic

...to selection_handler rather than selection_action. It's got nothing to do with the action buttons. Also, the signal works fine as an editor-defined static connection; completely skip the callback add/remove logic.

Koopa1018 committed:
De-static is_a_polygon_item() so it doesn't warn

I forgot Godot was annoying about treating statics as local fns :unamused:

Koopa1018 commented:
With the current state, we have a polygon editor which is basically usable. Edge selection is going to be a whole new job's worth of work, and isn't essential for the immediate term. What say we get this merged as-is and handle the other two checkboxes later?

Koopa1018 committed:
Move placed verts via drag, not click

Keeping this commit light, and there's STILL a bit to unpack! The gist of it is, placed verts now respond to left clicks on press, so it feels like dragging and dropping. Changed their mouse icon to make that clearer, too.

Implementing this required a new script for placed verts, so that right click (to delete) could still respond to release only. (Because each button is now specially handled, the button action mode is now fully ignored.)

Kind of awesome that I didn't have to modify polygon_editor or polygon_draw to end drag on mouse up--it just Already Did That.

Koopa1018 committed:
Use pointing-hand cursor during vertex delete

It feels *wrong* somehow to have the move icon while the vertex is being deleted.

Refactor fn to formally make room for handling other than main clicks.

Koopa1018 committed:
Put cursor in dragging state while dragging verts

Better than reverting to pointer state, I'd say.

Koopa1018 committed:
Hide add-vert button when mouse is near drag range

Helps avoid visual clutter, I think.

Koopa1018 committed:
Show the edited polygon in realtime

Interesting that the body pattern shifts on finalizing edits. Which I think is because the origin is changed when the polygon is totally rebuilt.... We're not even planning to save polygon origins, are we? I think I can fix that by just originating all polygons at (0,0)....(or just at snapped intervals?)

Includes an on/off switch just in case.

cruglet committed:
Fix add-vertex button not appearing on last edge

Original author: @cruglet
Original title: "Fix errors regarding out of index vertices/lines"

Also fixes a related bug where the display polygon would crash when verts were added to the last edge, due to out-of-range indexing.

Co-Authored-By: Cruglet <175854316+cruglet@users.noreply.github.com>

cruglet committed:
Update polygon display after each vertex added

Simple enough, fixes the issue where only the first two vertices and the last vertex (EDIT: of the display polygon -@koopa1018) would be the only ones to be rendered. Also, maybe we should make the terrain that is being modified slightly transparent? 🤔

Co-Authored-By: Cruglet <175854316+cruglet@users.noreply.github.com>

Koopa1018 committed:
Comment and cleanup last-edge fixes

Cleanup today includes respacing and underscoring added functions to standard, as well as adding an error assertion to safeguard against unintended behavior in (hopefully) a clear, readable manner.

cruglet committed:
Show polygon validity

Added functionality to show whether a polygon is valid or not, by making the outline red if any self-intersections are found within the polygon. Also checks for stacked vertices on top of each other. Will revisit later.

cruglet committed:
Terrain selection handler fix

Fixed issue where the selection handler would take priority and try to open the item properties menu for terrain. Also made sure that when the terrain is freed through the selection handler it is at least able to run drawable_polygon.end_edit before it exits the scene tree.

cruglet committed:
Reset valid state after terrain freed

Self explanatory, if the terrain was valid before a new one was made then the state would carry over

cruglet committed:
Update validity when removing vertex

Koopa1018 committed:
Refactor to standard, and for clarity

Space all functions properly, fix single-line if statements, and put comments in more sensible spots.

This seems like a good time to make the functions of PolygonValidator static, since PolygonValidator has no state anyway. Instancing a class fresh every time you need its functions is what I'd call code smell.

Koopa1018 committed:
Update invalidity marking during poly creation

Having the first line in invalid red is quite alarming and quite unnecessary; and it is, in fact, possible to have >3 verts during vertex creation, meaning validation *does* need to be done.

Since it's the same validation code as during dragging, let's just update both to use the same block of code.

Koopa1018 committed:
Merge adjacent verts that share a position

Why highlight verts in invalid positions when you could just...not allow invalid positions?

This took a lot of work to get working correctly :laughing:

(What about non-adjacent colocated verts? Simple. I believe those should not be considered invalid.)

Koopa1018 committed:
My bad, invalid 3-point shapes can exist

...when the shape is completely flat, all verts colinear. In that case, all edges intersect with at least one other edge.

Koopa1018 committed:
Refactor self-intersect detection to return edges

Or rather, first vertices of edges. Currently this is just used as a slower version of the same function, but my eventual plan is to use the extra data to highlight only the specific edges that intersect.

Koopa1018 committed:
Re-enable property menu crash for terrain

Terrain actually does need a property menu. Creating it is out of scope for this PR, but I'd rather have the crash than not--that way it's a constant reminder "implement the property menu, implement the..."
